### ISSUE TITLE
Problem in step 1 of SAIGE: fitting the null

### ISSUE THREAD

**Issue by dianacornejo** [weizhouUMICH/SAIGE#185]
Hi @weizhouUMICH 

I am running SAIGE for a binary trait with the UK Biobank data. Below is the command line I'm using with version SAIGE_0.38
```
Rscript ~/software/bin/step1_fitNULLGLMM.R \
    --plinkFile=plink_file \
    --phenoFile=phenotypic_file \
    --phenoCol=binary_trait \
    --covarColList=AGE,SEX \
    --sampleIDColinphenoFile=IID \
    --traitType=binary \
    --outputPrefix=output \
    --nThreads=20 \
    --LOCO=TRUE
```
And I'm getting the following error:

```
Error in X1 %*% firthfit$beta : non-conformable arguments
Calls: fitNULLGLMM -> <Anonymous>
In addition: Warning messages:
1: In glm.fit(x = c(-3495.24076423934, -6.07847527228668e-11, -6.07847527228668e-11,  :
  non-finite coefficients at iteration 13
2: glm.fit: algorithm did not converge
3: glm.fit: fitted probabilities numerically 0 or 1 occurred
4: In glm.fit(x = c(-3495.24076423934, -6.07847527228668e-11, -6.07847527228668e-11,  :
  non-finite coefficients at iteration 13
5: glm.fit: algorithm did not converge
6: glm.fit: fitted probabilities numerically 0 or 1 occurred
Execution halted
```
I appreciate any help in solving this issue!

Thanks

**Comment by weizhouUMICH** [comment]
Hi @dianacornejo, 

It seems there is the (nearly) perfect separation by some binary covariates. 
--minCovariateCount can be used to remove those covariates. The default value is -1. You may try a higher positive number, e.g. 10 or 50. 
This mean if a cell of the 2x2 table with the covariate and phenotype has a count number < minCovariateCount, the covariate will be removed. 
https://github.com/weizhouUMICH/SAIGE/blob/master/extdata/step1_fitNULLGLMM.R#L87 

Thanks,
Wei

**Comment by dianacornejo** [comment]
Hi @weizhouUMICH 
I set `--minCovariateCount` to 10, and now I'm getting this error

```
Error in if (iter == maxit | ((max(abs(delta)) <= xconv) & (all(abs(U.star[col.fit]) <  :
  missing value where TRUE/FALSE needed
Calls: fitNULLGLMM -> <Anonymous> -> fast.logistf.fit
In addition: Warning messages:
1: glm.fit: fitted probabilities numerically 0 or 1 occurred
2: glm.fit: fitted probabilities numerically 0 or 1 occurred
Execution halted
```
Any ideas on why could this be happening?

I appreciate your help a lot

**Comment by weizhouUMICH** [comment]
Hi @dianacornejo, 

How many covariates are left in the data set now? The Warning message suggests the perfect separation still exists. 

Thanks,
Wei

**Comment by dianacornejo** [comment]
Hi @weizhouUMICH,
I also tried running with no covariates and keep getting the same error. 
```
Error in if (iter == maxit | ((max(abs(delta)) <= xconv) & (all(abs(U.star[col.fit]) <  :
  missing value where TRUE/FALSE needed
Calls: fitNULLGLMM -> <Anonymous> -> fast.logistf.fit
In addition: Warning messages:
1: glm.fit: fitted probabilities numerically 0 or 1 occurred
2: glm.fit: fitted probabilities numerically 0 or 1 occurred
Execution halted
```
Thanks,
Diana

**Comment by weizhouUMICH** [comment]
Hi Diana, 

I see. Sorry I missed the glm.fit in the error message. I think the error is from the glm function in R to simply fit a logistic model on the data

` fit0 = glm(formula.new, data=data.new, family=binomial)`

Given that even without any covariate, the error persists, I think it is because the case number is too small compared to the control number. 
May I ask how many cases and controls there are in your data set?  

Thanks,
Wei

**Comment by dianacornejo** [comment]
Hi @weizhouUMICH 
Sure here is that information,
Number of cases: n=48625
Number of controls: n=290729
Thanks,
Diana

**Comment by weizhouUMICH** [comment]
Thanks!
Then the number of cases is not low at all. 
So even without any covariate, the error persists.
Could you please try this in R?  
`formula = as.formula(y~1)
fit0 = glm(formula, data=data, family=binomial)`

Also would you mind sharing the full log file?

Thanks,
Wei

**Comment by dianacornejo** [comment]
Hi @weizhouUMICH,
I followed your recommendation using this `formula = as.formula(y~1) fit0 = glm(formula, data=data, family=binomial)` in R I did not find any problems.
Please find attached the complete log file from the run without covariates.
[saige-step1-nocov-052020.log](https://github.com/weizhouUMICH/SAIGE/files/4658665/saige-step1-nocov-052020.log)
Thank you

**Comment by o147942** [comment]
> Thanks!
> Then the number of cases is not low at all.
> So even without any covariate, the error persists.
> Could you please try this in R?
> `formula = as.formula(y~1) fit0 = glm(formula, data=data, family=binomial)`
> 
> Also would you mind sharing the full log file?
> 
> Thanks,
> Wei

Hi, Wei.
I have the same error in the step1, but the number of cases is very low comparing to controls, case: 30 vs. control 3,000 .
In this case, I wonder if I cannot use this model.

**Comment by weizhouUMICH** [comment]
Hi @dianacornejo, 

Sorry for dropping the ball and thank you for sharing the log file. The log file shows the intercept estimate is really large from the glm function in R when there is no covariates: 5.716e+140. 
exp(intercept estimate) is supposed to be close to the proportion of cases in the data set. 
I guess something went wrong in the data set. But you mentipend that running the glm in R itself worked. Could you please share the coefficients estimated by glm in R 
`formula = as.formula(y~1) fit0 = glm(formula, data=data, family=binomial), (summary(fit0))$coefficients`

Also can you please try turn off the qr transformation --isCovariateTransform=FALSE to see if it makes difference? 

Thanks,
Wei



> Hi @weizhouUMICH,
> I followed your recommendation using this `formula = as.formula(y~1) fit0 = glm(formula, data=data, family=binomial)` in R I did not find any problems.
> Please find attached the complete log file from the run without covariates.
> [saige-step1-nocov-052020.log](https://github.com/weizhouUMICH/SAIGE/files/4658665/saige-step1-nocov-052020.log)
> Thank you

**Comment by weizhouUMICH** [comment]
Hi @o147942, 

Would you mind sharing the log file for the job with error? 

Thanks,
Wei


> > Thanks!
> > Then the number of cases is not low at all.
> > So even without any covariate, the error persists.
> > Could you please try this in R?
> > `formula = as.formula(y~1) fit0 = glm(formula, data=data, family=binomial)`
> > Also would you mind sharing the full log file?
> > Thanks,
> > Wei
> 
> Hi, Wei.
> I have the same error in the step1, but the number of cases is very low comparing to controls, case: 30 vs. control 3,000 .
> In this case, I wonder if I cannot use this model.

**Comment by xuefenfei712** [comment]
Hi Wei,
I use SAIGE today, but get the error below,
[1]   0 NaN
tau:  1 Inf 
tau0:  1 0.1 
Error in if (max(abs(tau - tau0)/(abs(tau) + abs(tau0) + tol)) < tol) break : 
  missing value where TRUE/FALSE needed
Calls: fitNULLGLMM -> system.time -> glmmkin.ai_PCG_Rcpp_Binary
In addition: Warning messages:
1: glm.fit: algorithm did not converge 
2: glm.fit: fitted probabilities numerically 0 or 1 occurred 
3: glm.fit: algorithm did not converge 
4: glm.fit: fitted probabilities numerically 0 or 1 occurred 
Timing stopped at: 665.9 15.09 169
Execution halted

thankyou!
Liu

**Comment by weizhouUMICH** [comment]
Sorry fo the late reply! We have just released a new version 1.0.0. It has computational efficiency improvements for both Step 1 and Step 2 for single-variant and set-based tests. We have created a new program github page https://github.com/saigegit/SAIGE with the documentation provided https://saigegit.github.io/SAIGE-doc/
The program will be maintained by multiple SAIGE developers there. The docker image has been updated. Please feel free to try the version 1.0.0 and report issues if any.

Thanks!
Wei
